src/main/java/com/couchbase/lite/replicator/PullerInternal.java

```diff
@@ -880,45 +877,87 @@ protected void stopGraceful() {
         new Thread(new Runnable() {
             @Override
             public void run() {
-
                 try {
-                    // stop things and possibly wait for them to stop ..
-                    if (batcher != null) {
-                        Log.d(Log.TAG_SYNC, "batcher.waitForPendingFutures()");
-                        // TODO: should we call batcher.flushAll(); here?
-                        batcher.waitForPendingFutures();
-                    }
-
-                    Log.d(Log.TAG_SYNC, "waitForPendingFutures()");
-                    waitForPendingFutures();
-
-                    if (downloadsToInsert != null) {
-                        Log.d(Log.TAG_SYNC, "downloadsToInsert.waitForPendingFutures()");
-                        // TODO: should we call downloadsToInsert.flushAll(); here?
-                        downloadsToInsert.waitForPendingFutures();
-                    }
+                    // wait for all tasks completed
+                    waitForAllTasksCompleted();
 
+                    // stop change tracker
                     if (changeTracker != null) {
                         Log.d(Log.TAG_SYNC, "stopping change tracker");
                         changeTracker.stop();
                         Log.d(Log.TAG_SYNC, "stopped change tracker");
                     }
-
-                }
-                catch (Exception e) {
+                } catch (Exception e) {
                     Log.e(Log.TAG_SYNC, "stopGraceful.run() had exception: %s", e);
                     e.printStackTrace();
-                }
-                finally {
+                } finally {
+                    // stop replicator immediate
                     triggerStopImmediate();
                 }
-
-                Log.e(Log.TAG_SYNC, "PullerInternal stopGraceful.run() finished");
+                Log.d(Log.TAG_SYNC, "PullerInternal stopGraceful.run() finished");
             }
         }).start();
     }
 
+    protected boolean waitingForPendingFutures = false;
+    protected Object lockWaitForPendingFutures = new Object();
+
     public void waitForPendingFutures() {
+        if (waitingForPendingFutures) {
+            return;
+        }
+        synchronized (lockWaitForPendingFutures) {
+            waitingForPendingFutures = true;
+
+            Log.d(Log.TAG_SYNC, "[PullerInternal.waitForPendingFutures()] STARTED - thread id: " + Thread.currentThread().getId());
+
+            try {
+                waitForAllTasksCompleted();
+            } catch (Exception e) {
+                Log.e(Log.TAG_SYNC, "Exception waiting for pending futures: %s", e);
+            } finally {
+                Log.d(Log.TAG_SYNC, "[waitForPendingFutures()] END - thread id: " + Thread.currentThread().getId());
+                fireTrigger(ReplicationTrigger.WAITING_FOR_CHANGES);
+                waitingForPendingFutures = false;
+            }
+        }
```

Comment:
From your comment on the previously closed PR, you would like to block so that there will be only one thread executing this code at a time. If I understand correctly, you don't need waitingForPendingFutures variable and the if statement above the synchronized block. Even put synchronized at the method level should achieve your goal. 

Currently there are 2 different scenarios that could happen depending on racing condition:
1. The "if (waitingForPendingFutures)" outside the synchronized block means there is no needs to execute this method  when one of thread is executing.
2. Some threads may get blocked at the synchronized block.